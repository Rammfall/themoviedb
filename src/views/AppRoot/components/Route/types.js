import PropTypes from 'prop-types'

export const propTypes = {
  path: PropTypes.string.isRequired,
  component: PropTypes.elementType.isRequired,
  exact: PropTypes.bool
}

export const defaultProps = {
  exact: false
}
