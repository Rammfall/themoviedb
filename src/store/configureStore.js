import { createLogicMiddleware } from 'redux-logic'
import { createStore, applyMiddleware } from 'redux'

import logics from './logics'
import rootReducer from './rootReducer'

const bindMiddleware = (middleware) => {
  if (process.env.NODE_ENV !== 'production') {
    // eslint-disable-next-line
    const { composeWithDevTools } = require('redux-devtools-extension')

    return composeWithDevTools(applyMiddleware(...middleware))
  }

  return applyMiddleware(...middleware)
}

const configureStore = (initialState = {}) => {
  // Add dependencies to pass them to logic functions
  const dependencies = {}

  const logicMiddleware = createLogicMiddleware(logics, dependencies)
  const store = createStore(rootReducer, initialState, bindMiddleware([logicMiddleware]))

  return store
}

export default configureStore
