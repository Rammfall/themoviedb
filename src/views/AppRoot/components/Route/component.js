import React from 'react'
import { Route, Redirect } from 'react-router-dom'
import PropTypes from 'prop-types'

import { propTypes, defaultProps } from './types'

const RouteComponent = ({
  component: Component,
  conditionRender,
  exact,
  path,
  redirectTo
}) => (
  <Route
    exact={exact}
    path={path}
  >
    {conditionRender ? <Component /> : <Redirect to={redirectTo} />}
  </Route>
)

RouteComponent.propTypes = {
  ...propTypes,
  redirectTo: PropTypes.string.isRequired
}

RouteComponent.defaultProps = {
  ...defaultProps
}

export default RouteComponent
