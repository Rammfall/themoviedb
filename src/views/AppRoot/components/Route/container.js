import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'

import { isLoggedInSelector } from 'Store/concepts/session/selectors'

import { defaultProps, propTypes } from './types'

import RouteComponent from './component'

class Route extends Component {
  get redirectTo() {
    const { privateRoute } = this.props

    return privateRoute ? '/login' : '/'
  }

  get isAccessAllowed() {
    const { privateRoute, isLoggedIn } = this.props

    return privateRoute ? isLoggedIn : !isLoggedIn
  }

  render() {
    const { exact, component, path } = this.props

    return (
      <RouteComponent
        exact={exact}
        path={path}
        component={component}
        redirectTo={this.redirectTo}
        isAccessAllowed={this.isAccessAllowed}
      />
    )
  }
}

Route.propTypes = {
  ...propTypes,
  privateRoute: PropTypes.bool,
  isLoggedIn: PropTypes.bool.isRequired
}

Route.defaultProps = {
  ...defaultProps,
  privateRoute: false
}

const mapStateToProps = (state) => ({
  isLoggedIn: isLoggedInSelector(state)
})

export { Route as RouteContainer }
export default connect(mapStateToProps)(Route)
