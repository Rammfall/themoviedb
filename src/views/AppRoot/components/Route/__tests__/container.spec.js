import React from 'react'
import { shallow } from 'enzyme'
import configureStore from 'redux-mock-store'

import Route from '../container'

jest.mock('Store/concepts/session/selectors', () => ({
  isLoggedInSelector: jest.fn(() => true)
}))

describe('Route', () => {
  describe('matches snapshot', () => {
    const mockStore = configureStore()()
    const Component = () => <div>Test</div>
    const defaultProps = {
      store: mockStore,
      path: '/',
      component: Component
    }

    it('with default props', () => {
      // eslint-disable-next-line react/jsx-props-no-spreading
      const container = shallow(<Route {...defaultProps} />)
        .dive()
        .dive()

      expect(container).toMatchSnapshot()
    })

    it('with privateRoute', () => {
      const privateProps = {
        ...defaultProps,
        privateRoute: true
      }
      // eslint-disable-next-line react/jsx-props-no-spreading
      const container = shallow(<Route {...privateProps} />)
        .dive()
        .dive()

      expect(container).toMatchSnapshot()
    })
  })
})
