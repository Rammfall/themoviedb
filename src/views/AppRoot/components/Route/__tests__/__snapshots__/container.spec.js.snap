// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Route matches snapshot with default props 1`] = `
<RouteComponent
  component={[Function]}
  exact={false}
  isAccessAllowed={false}
  path="/"
  redirectTo="/"
/>
`;

exports[`Route matches snapshot with privateRoute 1`] = `
<RouteComponent
  component={[Function]}
  exact={false}
  isAccessAllowed={true}
  path="/"
  redirectTo="/login"
/>
`;
